### ISSUE TITLE
[HOLD for payment 2023-08-16] [$1000] Web - Personal Details - US States dropdown is shown even when the country is not set to US

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#22018]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging dot on web chrome and login with a NEW account (imp step)
2. Settings > Personal information > Address
3. Make sure it's empty at first
4. If you click on state dropdown, notice the US states are shown.
5. But if you notice the Country is still set to null. Thus, the US states are shown on the State dropdown even when the Country is not set to US.

## Expected Result:
Country dropdown should be at the top. We should try to auto-detect the country, and if we can't, we should default country to US.

## Actual Result:
Country dropdown is towards the bottom of the form and the US State dropdown is shown even when the country is not set to US

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.35-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/e301954a-989b-4359-a100-56233343c2df

https://github.com/Expensify/App/assets/93399543/be16091b-05ab-4a27-9929-91c3108df0a9

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1688188803552689

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0122bc9c20db346e72</li>
        <li>Upwork Job ID: 1678797415883694080</li>
        2023-07-18<li>Automatic offers: </li>
<ul>
<li>zaitsev1 | Contributor | 25773784</li>
<li>priya-zha | Reporter | 25773788</li>
</ul></li>
    </ul>
</details>

**Comment by StevenKKC** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
US States dropdown is shown even when the country is not set to US in Personal Details - Home Address page.

### What is the root cause of that problem?
US States dropdown is displayed when `isUSAForm` is set, and `isUSAForm` is set to true when selected country ISO code is USA.
https://github.com/Expensify/App/blob/c0ece67e0edcb725d8f72cf57cd40d98854c552c/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L172-L177
When first open Home Address page, `countryISO` is retrived from personal details. And when user change country, `countryISO` is set to corresponding country code.
https://github.com/Expensify/App/blob/c0ece67e0edcb725d8f72cf57cd40d98854c552c/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L65
https://github.com/Expensify/App/blob/c0ece67e0edcb725d8f72cf57cd40d98854c552c/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L197-L201

But default country code is not set to `countryISO`.

### What changes do you think we should make in order to solve the problem?
Country should be set to `countryISO` as below.
```js
	<CountryPicker
		inputID="country"
		onValueChange={setCountryISO}
-		defaultValue={PersonalDetails.getCountryISO(address.country)}
+		defaultValue={countryISO}
	/>

```

### What alternative solutions did you explore? (Optional)
None.

**Comment by sakluger** [comment]
I agree that this is a bug for anyone outside of the US. I think that instead of making the State dropdown blank, we should instead move the country dropdown up to the top so that it's the first selection. cc @Expensify/design do you agree?

**Comment by shawnborton** [comment]
Yeah, I feel like we've chatted about this one before. But I could see having Country be the first dropdown, and we should try to auto-fill country if we an based on where the user is logging in from? 

cc @kevinksullivan @JmillsExpensify @anmurali @joekaufmanexpensify as I feel like we've talked about this one as it impacts Wallet/bank account flows too.

**Comment by sakluger** [comment]
Not overdue, just waiting for feedback from the team.

**Comment by joekaufmanexpensify** [comment]
Yeah, I like trying to auto-fill the country based on where the user logs in from. If we move country to the top, we'd still then just show the US states in the state field prior to a country being selected?

**Comment by shawnborton** [comment]
I guess so, yeah. Or alternatively we try to auto-detect country but if we can't grab a country, we just default to USA.

**Comment by joekaufmanexpensify** [comment]
Cool cool, that makes sense to me.

**Comment by sakluger** [comment]
I like defaulting to US if we can't detect the country.

**Comment by sakluger** [comment]
Updated the Expected and Actual Results in the OP, adding external label!

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Web - Personal Details - US States dropdown is shown even when the country is not set to US
### What is the root cause of that problem?

Since Auto-Detecting country not handled and the state that calculate's the user's country is not used into the input component here.
https://github.com/Expensify/App/blob/17d1c3dc1f16428802b151ac069db43f27a93aad/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L205

### What changes do you think we should make in order to solve the problem?
Even we do all things correct the value is only going to be "US" which is default one. 

To tackle this we should calculate user's country from his timezone using code like below

```js

function getCountryFromTimeZone(timezone) {
    const allCountries = CONST.ALL_COUNTRIES;

    // zones from `moment-timezone`
    const timezoneCountries = lodashGet(zones, `${timezone}.countries`);
    if (timezoneCountries.length === 0) {
        return;
    }
    // Multiple countries could have same timezone picking first country and countries is an array, picking `0`th index here.
    const countryCode = timezoneCountries[0];
    return countryCode;
}

function getUserCountryFromTimezone() {
    const timezone = getCurrentTimezone();
    const country = getCountryFromTimeZone(timezone.selected);
    return country;
}

// Set OnyxKey here and use it in the component
Onyx.set(ONYXKEYS.COUNTRY_ISO, getUserCountryFromTimezone());

```

We should use the Onyx key we've set above in the component here

https://github.com/Expensify/App/blob/17d1c3dc1f16428802b151ac069db43f27a93aad/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L218

We should re-order the fields in this order

* Country
* State
* City
** Other fields.


### What alternative solutions did you explore? (Optional)

NA

**Comment by sobitneupane** [comment]
```js     
// Multiple countries could have same timezone picking first country and countries is an array, picking `0`th index here.
    const countryCode = timezoneCountries[0];
    return countryCode;
```

@b4s36t4 Above code shows that if multiple countries have the same timezone, than we have a chance of showing a wrong country which doesn't sound like a right thing to do. Can we have some reliable way of detecting a country?

**Comment by b4s36t4** [comment]
We can do that but it includes an API call to some provider or open APIs using the longitude and latitude of the user's location. Maybe like below but this also includes user should allow GPS permission to the website.
<img width="584" alt="Screenshot 2023-07-13 at 2 34 22 PM" src="https://github.com/Expensify/App/assets/59088937/71749692-5781-4975-8a25-05ba5522e28f">

**Comment by b4s36t4** [comment]
for mobile we can use `react-native-device-info` package, but for web it might get complicated.

**Comment by b4s36t4** [comment]
@sobitneupane I feel the above solution is ok because most of the time we get distinct results. In the case of not able to find any country we're setting default as `US` right so I feel this comes under the same situation. 

If you don't want to make things harder for the user we could use user's `locale` if we have more than 1 country based on timezone we pick the country from `locale` maybe user `navigator.language`.

**Comment by sobitneupane** [comment]
>  I feel the above solution is ok because most of the time we get distinct results. In the case of not able to find any country we're setting default as US right so I feel this comes under the same situation.

I agree with you. But there will be the cases where we will be showing the wrong country which doesn't sound like a good plan.

**Comment by b4s36t4** [comment]
Yes, I feel same. I'm thinking that maybe if the timezone countries are more than 1 we get the user's locale and fill out the filed based on the value. This can help in most of the cases.

**Comment by zaitsev1** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Country dropdown to be located at the top of the form, and we aim to implement an auto-detection mechanism to identify the user's country. In cases where the auto-detection fails, the default country should be set to the United States

### What is the root cause of that problem?
1. We do not determine where the device is located.

2. by default in the countryISO state we have US set if the user has no saved address data. 
https://github.com/Expensify/App/blob/17d1c3dc1f16428802b151ac069db43f27a93aad/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L65
but CountryPicker has the wrong defaultValue, so we see the US states, but we don't see the US country selected.
https://github.com/Expensify/App/blob/17d1c3dc1f16428802b151ac069db43f27a93aad/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L202-L205

### What changes do you think we should make in order to solve the problem?
I propose to determine the user's country by his IP address, for this we can use one of the geolocation API that finds a user's country.

Example: 
1. https://country.is - free to use. No API keys or usage quotas.
2. https://ip.nf free to use. No API keys or usage quotas.
4. https://geolocation-db.com/json - It seems free.
3. https://ipinfo.io - paid
We can find other alternatives

If you prefer not to use the hosted service, run your own private server with [Docker](https://hub.docker.com/r/hakanensari/country).
Country.is uses geolocation data provided by [Cloudfare](https://support.cloudflare.com/hc/en-us/articles/200168236-Configuring-IP-geolocation) and [MaxMind](https://www.maxmind.com/).

Code:
```javascript
const setCountryCodeByTimeZone = () => {
    const deviceTimeZone = moment.tz.guess(); // automatic device timezone detection
    const countryCodeTimeZone = moment.tz.zone(deviceTimeZone).countries()[0];

    setCountryISO(countryCodeTimeZone);
};

useEffect(() => {
    if (countryISO) {
        return; // if the user already has a saved country then we will not recognize through ip
    }

    try {
        fetch(`https://ip.nf/me.json`)
            .then((response) => response.json())
            .then((data) => {
                const iso = PersonalDetails.getCountryISO(data.ip.country);

                if (iso) {
                    setCountryISO(iso);
                    return;
                }

                setCountryISO(CONST.COUNTRY.US);
            });
    } catch (error) {
        // if geolocation API is not available, recognize it via timezone
        setCountryCodeByTimeZone();
        console.error(error);
    }
}, [countryISO]);
``` 

### **Video of how it works** 
https://cln.sh/2kT1V8jLfczmDxdH3tns

### What alternative solutions did you explore? (Optional)
N/A

**Comment by zaitsev1** [comment]
Contributor details
Your Expensify account email: developer.zaitsev@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01790892be24aec889

**Comment by StevenKKC** [comment]
`countryCode` is returned by the `ReconnectApp` API based on the IP address. Now `countryCode` is used for generate phone number, but I think it is possible using `countryCode`.

**Comment by b4s36t4** [comment]
I think yes @StevenKKC solution works but with countryCode we don't have any mapping locally for that we can use the following method.

We can directly use this inside the https://github.com/Expensify/App/blob/17d1c3dc1f16428802b151ac069db43f27a93aad/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L63 and use it inside the component.

<img width="600" alt="Screenshot 2023-07-14 at 10 55 30 PM" src="https://github.com/Expensify/App/assets/59088937/869e71c5-fca2-487d-a33f-4376c07f860e">

cc: @sobitneupane

**Comment by b4s36t4** [comment]
@sobitneupane Any update on proposals?

**Comment by sakluger** [comment]
@sobitneupane mind taking another look at the updated proposals?

**Comment by b4s36t4** [comment]
@sobitneupane Updated the proposal https://github.com/Expensify/App/issues/22018#issuecomment-1631385840.

**Comment by b4s36t4** [comment]
@sobitneupane  let me know what you think about the proposal.

**Comment by zaitsev1** [comment]
@shawnborton Phone number codes are not unique to a single country. Some countries or regions may share the same or similar phone number codes, leading to ambiguity and confusion in identifying the correct country.

Example of identification by country code number:

https://github.com/Expensify/App/assets/32131367/886f50a2-593a-4bc4-85e3-e631af7a178d

You may note that it defines Canada as the United States, and there are many such examples:

<details><summary>Details</summary>
<p>

United States and Canada (+1) - As mentioned earlier, these countries share a common country code.

Anguilla, Antigua and Barbuda (+1) - Just like the US and Canada, they use the +1-country code.

Virgin Islands (British) (+1) - Also share a country code with the US, Canada, and other regions.

St. Maarten / St. Martin (+1) - The Caribbean Island shared between France and the Netherlands also uses the country code +1.

Malaysia (+60) and Brunei (+673) - These two countries share the +60-country code, which provides a common telephone system.

Australia (+61) and Norfolk Island (+672) - Norfolk Island, which is a territory of Australia, also shares a country code with Australia.

Guam (USA) (+1) and Northern Mariana Islands (USA) (+1) - Both US territories in the Pacific Ocean have a +1-country code.

Cayman Islands (+1) - This British Overseas Territory also uses the +1-country code, the same as the United States and Canada.

Bermuda (+1) - Another British Overseas Territory with a common country code of +1.

Palau (+680) and Japan (+81) - Palau and Japan share the prefix code +81.

China (+86) and Hong Kong (+852) - Hong Kong has its own prefix code +852, but is in the same country as China, which has a country code of +86.

France (+33) and Monaco (+377) - Monaco has its own country code +377, but it is located inside the territory of France, which has the +33-country code.

England (+44) and the Isle of Man (+44) - The Isle of Man, located between England and Ireland, also shares the country code +44 with England.

Portugal (+351) and Madeira (+351) - Madeira, an archipelago in northwestern Portugal, has a country code of +351, the same as mainland Portugal.

</p>
</details> 

Instead of relying solely on phone number codes, a more robust approach for country recognition at the frontend would involve utilizing geolocation APIs. These methods can provide more accurate and reliable results and improve the overall user experience by tailoring the frontend based on the user's actual location and preferences.

**Comment by sobitneupane** [comment]
> Instead of relying solely on phone number codes, a more robust approach for country recognition at the frontend would involve utilizing geolocation APIs. These methods can provide more accurate and reliable results and improve the overall user experience by tailoring the frontend based on the user's actual location and preferences.

Which API would you suggest? And will it require user to give permissions? We can start discussion in slack if you put forward a proposal.

**Comment by sobitneupane** [comment]
Thanks @zaitsev1 for the [proposal](https://github.com/Expensify/App/issues/22018#issuecomment-1642503245).

> One alternative solution could be modifying the backend to return the user's country based on their IP address, similar to how it currently returns the phone country code. However, implementing this on the backend might require additional resources and changes to existing systems.

I prefer the alternative proposal. Since we are already returning phone country code. I don't think it will be a difficult to do the same for country.

So, the solution of this issue involves change in backend to return user's country which can be later used in frontend.

🎀 👀 🎀 C+ reviewed

**Comment by jasperhuangg** [comment]
> Thanks @zaitsev1 for the [proposal](https://github.com/Expensify/App/issues/22018#issuecomment-1642503245).
> 
> > One alternative solution could be modifying the backend to return the user's country based on their IP address, similar to how it currently returns the phone country code. However, implementing this on the backend might require additional resources and changes to existing systems.
> 
> I prefer the alternative proposal. Since we are already returning phone country code. I don't think it will be a difficult to do the same for country.
> 
> So, the solution of this issue involves change in backend to return user's country which can be later used in frontend.
> 
> 🎀 👀 🎀 C+ reviewed

I can get a PR up that does this, we'll HOLD this on that for now, once it's out I can assign @zaitsev1, thanks!

**Comment by sakluger** [comment]
@JmillsExpensify mentioned in Slack that we already set a default output currency for all accounts when they sign-up, based on IP. Can we default the country based on the output currency that is associated with the account? cc @jasperhuangg @zaitsev1

**Comment by zaitsev1** [comment]
@sakluger Both phone number codes and currency are not always unique to a single country.

One prominent example is the Euro (EUR) is used as the official currency in 19 European Union member countries. Consequently, these countries share the same currency code (EUR), despite being separate sovereign nations.

Additionally, some countries and territories outside the Eurozone also use the Euro as their official or de facto currency, but they may have a different currency code or none at all. For example, Andorra and Monaco both use the Euro but are not part of the Eurozone and have their own currency codes (ADP and MCF, respectively). Other regions, like Kosovo and Montenegro, also use the Euro without having official codes assigned.

**Comment by jasperhuangg** [comment]
@zaitsev1 Agree, in this PR https://github.com/Expensify/Web-Expensify/pull/38284 we're now sending back the user's country from their IP in a new Onyx key `country`. Feel free to use that to implement your solution!

**Comment by zaitsev1** [comment]
@jasperhuangg 
Sorry for the long answer, we have different time zones

in the main branch I did not find the `country` in `ONYXKEYS.js `
Only `COUNTRY_CODE`, which returns the phone code

This repository seems to be private:
https://github.com/Expensify/Web-Expensify/pull/38284

**Comment by jasperhuangg** [comment]
@zaitsev1 You're going to have to add the key yourself to ONYXKEYS. It's being sent back and stored in Onyx from our API right now. That's what the private repo PR was doing.

Open the Inspector > Application > IndexedDB > OnyxDB > keyvaluepairs and you'll see that it's there:

<img width="714" alt="Screenshot 2023-07-28 at 3 44 58 PM" src="https://github.com/Expensify/App/assets/31285285/8e35d67a-9c30-47cb-a2f1-b6f9d8a7e9aa">

**Comment by zaitsev1** [comment]
Thanks, @jasperhuangg, I already realized that, and already preparing a PR.

I just thought it should have been in ONYXKEYS.js already

My fault for not realizing it earlier.

**Comment by zaitsev1** [comment]
@sobitneupane @jasperhuangg PR is ready to review https://github.com/Expensify/App/pull/23841 🚀

**Comment by sobitneupane** [comment]
@sakluger Looks like jasperhaagg is OOO, can we assign a new CME?

**Comment by zaitsev1** [comment]
I would like to note that @jasperhuangg was absent when I made the PR, so the time of approve pr was delayed

**Comment by jasperhuangg** [comment]
Yes absolutely @zaitsev1, thanks for noting that.

Let's go off of when the C+ submitted their review as the time the approval was made: https://github.com/Expensify/App/pull/23841#pullrequestreview-1554466076

**Comment by sakluger** [comment]
I reviewed the timing and I agree that this one qualifies for the bonus.

I completed payment for @priya-zha ($250) and @zaitsev1 ($1500) via upwork.

@sobitneupane can you please complete the BugZero checklist and then request payment via NewDot? Thanks!

**Comment by zaitsev1** [comment]
@sakluger I have successfully received the payment. 
I extend my gratitude to all involved.

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

Rather than a bug it was an intended behaviour back then. We have improved the behavior to detect country.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

Log into an account that does not have a saved address (this step is crucial to replicate the scenario).
Navigate to "Settings" and access the "Personal Information" section.
Click on "Home Address.”
Verify that country field in the form is auto selected and shows the user's country.

**Comment by sobitneupane** [comment]
Requested payment on newDot

https://github.com/Expensify/App/issues/22018#issuecomment-1681619052

**Comment by sakluger** [comment]
Thanks!

Summarizing payouts for this issue:
- Issue reporter: @priya-zha $250 (Paid via Upwork)
- Contributor:  @zaitsev1 $1500 (Paid via Upwork)
- Contributor+: @sobitneupane $1500 (Requested via manual request)

Upwork job is linked [here](https://www.upwork.com/jobs/~0122bc9c20db346e72)

**Comment by sakluger** [comment]
@JmillsExpensify @anmurali assigning so that you can approve C+ payment.

**Comment by anmurali** [comment]
@JmillsExpensify is handling these, so I am unassigning myself.

**Comment by sakluger** [comment]
@JmillsExpensify told me that this is already in his list to pay so I'm going to close the issue.

**Comment by JmillsExpensify** [comment]
Reviewed the details for @sobitneupane. $1,500 approved for payment in NewDot based on BZ summary.
